File: .agents/skills/upgrade-harness/scripts/harness_upgrade.py
#!/usr/bin/env python3
"""为下游 Harness 工程升级生成三方计划并安全更新既有受管文件。"""

from __future__ import annotations

import argparse
import json
from pathlib import Path
import sys

from harness_upgrade_core import UpgradeError, build_plan, write_new_plan
from harness_upgrade_mutation import apply_plan, record_lock


def add_shared_arguments(parser: argparse.ArgumentParser) -> None:
    """为 plan 子命令添加候选、目标、清单和来源锁参数。"""

    parser.add_argument("--source-root", type=Path, required=True)
    parser.add_argument("--source-version", required=True)
    parser.add_argument("--source-commit", required=True)
    parser.add_argument("--candidate-root", type=Path, required=True)
    parser.add_argument("--target-root", type=Path, required=True)
    parser.add_argument("--ownership", type=Path, required=True)
    parser.add_argument("--lock", type=Path, required=True)


def build_parser() -> argparse.ArgumentParser:
    """构造稳定的非交互命令行接口。"""

    parser = argparse.ArgumentParser(description=__doc__)
    commands = parser.add_subparsers(dest="command", required=True)
    plan = commands.add_parser("plan", help="Generate a read-only three-way update plan")
    add_shared_arguments(plan)
    plan.add_argument("--output", type=Path)

    apply_command = commands.add_parser(
        "apply",
        help="Apply approved updates to existing unchanged managed files",
    )
    apply_command.add_argument("--plan", type=Path, required=True)
    apply_command.add_argument("--approval", required=True)
    apply_command.add_argument("--path", required=True)

    record = commands.add_parser("record", help="Record the reviewed Harness baseline")
    record.add_argument("--plan", type=Path, required=True)
    record.add_argument("--source-version", required=True)
    record.add_argument("--source-commit", required=True)
    record.add_argument("--approval", required=True)
    record.add_argument("--bootstrap", action="store_true")
    record.add_argument("--resolved-manual", action="append", default=[])
    return parser


def main() -> int:
    """解析命令、输出结构化结果，并以非零状态暴露安全阻断。"""

    parser = build_parser()
    args = parser.parse_args()
    try:
        if args.command == "plan":
            result = build_plan(
                args.source_root,
                args.source_version,
                args.source_commit,
                args.candidate_root,
                args.target_root,
                args.ownership,
                args.lock,
            )
            if args.output:
                write_new_plan(
                    args.output,
                    result,
                    (
                        Path(result["candidate_root"]),
                        Path(result["target_root"]),
                    ),
                )
            print(json.dumps(result, ensure_ascii=False, sort_keys=True))
            return 2 if result["blocked"] else 0
        if args.command == "apply":
            result = apply_plan(args.plan, args.approval, args.path)
        else:
            result = record_lock(args)
        print(json.dumps(result, ensure_ascii=False, sort_keys=True))
        return 0
    except (OSError, UpgradeError, TypeError, ValueError) as exc:
        print(
            json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False),
            file=sys.stderr,
        )
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: .agents/skills/desktop-upgrade-harness/scripts/harness_upgrade.mjs
#!/usr/bin/env node
/** 为下游 Harness 工程升级生成三方计划并安全更新既有受管文件。 */

import path from "node:path";
import process from "node:process";
import { pathToFileURL } from "node:url";
import { parseArgs } from "node:util";
import { buildPlan, stableJson, writeNewPlan } from "./harness_upgrade_core.mjs";
import { applyPlan } from "./harness_upgrade_mutation.mjs";
import { recordLock } from "./harness_upgrade_record.mjs";
import { UpgradeError } from "./harness_upgrade_safety.mjs";

function parseCommand(argv) {
  const command = argv[0];
  if (!["plan", "apply", "record"].includes(command)) throw new UpgradeError("必须提供 plan、apply 或 record 子命令");
  let options;
  if (command === "plan") options = Object.fromEntries(["source-root", "source-version", "source-commit", "candidate-root", "target-root", "ownership", "lock", "output"].map((name) => [name, { type: "string" }]));
  else if (command === "apply") options = { plan: { type: "string" }, approval: { type: "string" }, path: { type: "string" } };
  else options = { plan: { type: "string" }, "source-version": { type: "string" }, "source-commit": { type: "string" }, approval: { type: "string" }, bootstrap: { type: "boolean", default: false }, "resolved-manual": { type: "string", multiple: true, default: [] } };
  let values;
  try { ({ values } = parseArgs({ args: argv.slice(1), options, strict: true })); } catch (error) { throw new UpgradeError(error.message); }
  const required = command === "plan" ? ["source-root", "source-version", "source-commit", "candidate-root", "target-root", "ownership", "lock"] : (command === "apply" ? ["plan", "approval", "path"] : ["plan", "source-version", "source-commit", "approval"]);
  if (required.some((name) => !values[name])) throw new UpgradeError("缺少必需命令参数");
  return { command, values };
}

export function execute(argv) {
  const { command, values } = parseCommand(argv);
  if (command === "plan") {
    const result = buildPlan(values["source-root"], values["source-version"], values["source-commit"], values["candidate-root"], values["target-root"], values.ownership, values.lock);
    if (values.output) writeNewPlan(values.output, result, [result.candidate_root, result.target_root]);
    return { result, exitCode: result.blocked ? 2 : 0 };
  }
  if (command === "apply") return { result: applyPlan(values.plan, values.approval, values.path), exitCode: 0 };
  return { result: recordLock({ plan: values.plan, sourceVersion: values["source-version"], sourceCommit: values["source-commit"], approval: values.approval, bootstrap: values.bootstrap, resolvedManual: values["resolved-manual"] }), exitCode: 0 };
}

if (process.argv[1] && import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href) {
  try { const { result, exitCode } = execute(process.argv.slice(2)); process.stdout.write(`${stableJson(result)}\n`); process.exitCode = exitCode; }
  catch (error) { process.stderr.write(`${stableJson({ ok: false, error: error.message })}\n`); process.exitCode = 2; }
}
